File: operators/communication/all_to_all.py
# Copyright (c) 2024 The Core Team
#
# Licensed under the Apache License, Version 2.0
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import torch
from typing import List

def all_to_all(tensors: List[torch.Tensor], dim=0) -> List[List[torch.Tensor]]:
    """
    Simulate an all-to-all operation where each process sends and receives a segment of data to and from every other process.
    Each tensor in the input list represents the complete data held by one process, which needs to be divided equally among all processes.
    """
    if not tensors:
        raise ValueError("The tensor list cannot be empty.")

    num_processes = len(tensors)
    # Split each tensor into parts, one for each process
    segments = [torch.chunk(tensor, num_processes, dim=dim) for tensor in tensors]

    # Distribute segments to each process
    result = [[] for _ in range(num_processes)]
    for i in range(num_processes):
        for j in range(num_processes):
            result[j].append(segments[i][j].to(tensors[j].device))

    return result

File: src/communication/all_to_all.cpp
/*
 * Copyright (c) 2024 The Core team
 *
 * Licensed under the Apache License, Version 2.0;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "communication/all_to_all.h"

#include <torch/csrc/autograd/generated/variable_factories.h>

#include <limits>

#include "communication/communication_impl.h"
#include "logger.h"
#include "tensor_impl.h"
#include "threading/scheduler_impl.h"
#include "threading/task_impl.h"

namespace cs::communication {
void AllToAll::run(const Scheduler &scheduler, const Comm &comm,
                   const std::vector<Tensor> &tensorReceive,
                   const std::vector<ReadOnlyTensor> &tensorSend) {
  struct Impl : Task::Impl {
    const Comm comm;

    explicit Impl(std::vector<Tensor> output /* tensors */,
                  std::vector<ReadOnlyTensor> input /* tensors */, Comm comm)
        : Task::Impl{std::move(output), std::move(input), kComm, kNccl},
          comm{std::move(comm)} {}
    void operator()() const override {
      std::vector<at::Tensor> vReceive;
      vReceive.reserve(output().size());
      for (const auto &t : output()) {
        CS_ASSERT_TRUE(t.impl()->tensor().device().type() == at::kCUDA,
                       "NCCL backend only support CUDA GPUs");
        vReceive.push_back(t.impl()->tensor());
      }

      std::vector<at::Tensor> vSend;
      vSend.reserve(input().size());
      for (const auto &t : input()) {
        CS_ASSERT_TRUE(t.impl()->tensor().device().type() == at::kCUDA,
                       "NCCL backend only support CUDA GPUs");
        vSend.push_back(t.impl()->tensor());
      }

      comm.impl()->backend()->alltoall(vReceive, vSend)->wait();
    }
    [[nodiscard]] const char *name() const override {
      return "cs::communication::AllToAll::run";
    }
  };

  scheduler.impl()->submit(
      Task{std::make_shared<Impl>(Impl{tensorReceive, tensorSend, comm})});
}
}  // namespace cs::communication
